lib/Importer/Caffe2ModelLoader.cpp

```diff
@@ -942,28 +942,22 @@ llvm::Error Caffe2ModelLoader::loadOperator(const caffe2::OperatorDef &op) {
     Node *node;
     if (isFused) {
       // There is no specific fused quantized type in Caffe2, so we will load
-      // Int8QTy. We then change it from Int8QTy to Int8FusedQTy here if
+      // Int8QTy. We then change it from Int8QTy to UInt8FusedQTy here if
       // necessary -- another user could have already changed it.
-      if (dataC->getElementType() != ElemKind::Int8FusedQTy) {
+      if (dataC->getElementType() != ElemKind::UInt8FusedQTy) {
         RETURN_ERR_IF_NOT(dataC->getElementType() == ElemKind::Int8QTy,
                           "Data must be Int8QTy.");
         // Use dummy 0.0/0 as scale/offset, since the actual scales/offsets are
         // fused inline with the data.
-        TypeRef fusedTy = G_.getParent()->uniqueType(ElemKind::Int8FusedQTy,
+        TypeRef fusedTy = G_.getParent()->uniqueType(ElemKind::UInt8FusedQTy,
                                                      dataC->dims(), 0.0, 0);
         dataC->setType(Storage::OutputIdx, fusedTy);
-      }
 
-      // Caffe2 stores offsets as floats, whereas we want to use int32_t.
-      char *dataBasePtr = dataC->getPayload().getUnsafePtr();
-      const size_t width = dataC->dims()[1];
-      for (size_t i = 0, e = dataC->dims()[0]; i < e; ++i) {
-        // Must memcpy to the stack and back to avoid misaligned addresses.
-        char *currRowOffsetPtr = dataBasePtr + (i + 1) * width - 4;
-        float fOffset;
-        memcpy(&fOffset, currRowOffsetPtr, 4);
-        int32_t iOffset = static_cast<int32_t>(fOffset);
-        memcpy(currRowOffsetPtr, &iOffset, 4);
+        // We also need to update the data to be unsigned instead of signed.
+        auto dataCH = dataC->getHandle<int8_t>();
+        for (size_t i = 0, e = dataCH.size(); i < e; i++) {
+          dataCH.raw(i) = (uint8_t)dataCH.raw(i) + OFFSETSHIFT;
```

Comment:
I'm a bit worried about potential UB here since dataCH.raw(i) is signed, and we might be trying to store an out-of-range uint8_t in it.

Reply:
This was actually incorrect -- the Tensor should have been `uint8_t` (I changed the Constant but not the payload Tensor). And it wasn't an issue because when we generate the runtime bundle the type is just taken from the Constant and not the Tensor. Anyway, I fixed it and added in a verification check in `Function::verify()`.